### PR TITLE
Allowed super-generic predicates in FluentWait to avoid having to cast

### PR DESCRIPTION
**Pull request by skestle**
FluentWaits or expected predicates.

Have signed CLA

### CODE REVIEW

**Comment by skestle** [comment]
Is there something holding this up? Perhaps you'd like me to go through ExpectedConditions and widen the generic types?

**Comment by andreastt** [comment]
Maybe @shs96c would like to take a look?

**Comment by barancev** [comment]
Is there a real-life example where it can be used? What kind of casting do you avoid?

**Comment by skestle** [comment]
TL:DR; I wrote the full text before I checked the tests - they seem clear to me: Why would I write predicates against `WebDriver` when they work just as well against `WebElement`? Existing `WebDriverWait`s cannot work with `SearchContext` predicates.

**Full Answers**

The following example highlights the multiple issues with the current Expectations in the selenium code base - my particular code relies on highly encapsulated and cohesive structures that would distract from the issue.

    WebElement containerTable;
    Utils.newFluentWait((SearchContext) containerTable).until(ExpectedConditions.visibilityOfElementLocated(By.xpath("tr[1]"));

Where `visibilityOfElementLocated` would of course return a `Predicate<SearchContext>` which is implemented by a heap of objects, and `newFluentWait` calls `new FluentWait<T>(e)`

Having to pass `WebDriver` around is bad design and is not helpful searching since it is effectively the `body` scope.

So, "avoiding casts" is one way of looking at it when writing utility classes. The reverse problem is having to implement my `SearchContextWebDriverHack` class:

	public class SearchContextWebDriverHack implements WebDriver {

	private final SearchContext searchContext;

	public SearchContextWebDriverHack(final SearchContext searchContext) {
		this.searchContext = searchContext;
	}
	...

Which results in the initial code becoming something like:

    new WebDriverWait(new SearchContextWebDriverHack(containerTable)).until(ExpectedConditions.visibilityOfElementLocated(By.xpath("tr[1]"));

**Comment by skestle** [comment]
Or, for a more succinct "real world" example, update 'ExpectedConditions' to return `Predicate<SearchContext>` where possible (`*located*` methods), and see the carnage.

**Comment by skestle** [comment]
Of course, these "concrete" requirements are just the beginning (establishing something useful even with basic selenium). 
Given composable/encapsulated objects that interact with different parts of pages etc, these super generics allow all sorts of nice DSL like behaviour.

I know, OO right? Wow.

**Comment by skestle** [comment]
Can we acknowledge that this is equivalent to the change made in https://github.com/SeleniumHQ/selenium/commit/59b58915d7889b13752b03b0aa5e0a1c435c2681 for the `until(Function<? super T, V>)` method (that was previously `Function<T,V>`).

Please. Pushing developers to presume `WebDriver` all over the place is not good.

**Comment by andreastt** [comment]
@jleyba

**Comment by skestle** [comment]
Fixed by b2aa9fd534f7afbcba319231bb4bce85f825ef09
